test/decl/func/dynamic_self.swift

```diff
@@ -412,4 +412,88 @@ class SelfOperator {
 func useSelfOperator() {
   let s = SelfOperator()
   _ = s + s
-}
\ No newline at end of file
+}
+
+// These references to Self are now possible
+
+class A<T> {
+  let b: Int
+  required init(a: Int) {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    b = a
+  }
+  static func z() {
+    print("\(Self.self).\(#function)")
+  }
+  class func y() {
+    print("\(Self.self).\(#function)")
+    Self.z()
+  }
+  func x() -> A? {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    return Self.init(a: 77)
+  }
+}
+
+class B: A<Int> {
+  let a: Int
+  required convenience init(a: Int) {
+    print("\(Self.self).\(#function)")
+    self.init()
+  }
+  init() {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    a = 99
+    super.init(a: 88)
+  }
+  override class func y() {
+    print("override \(Self.self).\(#function)")
+  }
+}
+
+struct S2 {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
+    Self.x()
+    return Self.init() as Self
+  }
+}
+
+struct S3<T> {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
+    Self.x()
+    return Self.init() as? Self
+    // expected-error@-1 {{cannot convert return expression of type 'S3<T>?' to return type 'S3<T>?'}}
+    // expected-error@-2 {{generic type cannot be used in this context}}
+    // expected-error@-3 {{generic type cannot be used in this context}}
+  }
+}
+
+enum E {
+  static func f() {
+    print("f()")
+  }
+  case e
+  func h(h: Self) -> Self {
+    Self.f()
+    return .e
+  }
+}
```

Comment:
You should also test nested types and extensions

Reply:
Restructures tests to cover extensions and nested types.